plugins/storage/fileshare/storagefsvm/src/main/java/org/apache/cloudstack/storage/fileshare/lifecycle/StorageFsVmFileShareLifeCycle.java

```diff
@@ -0,0 +1,313 @@
+import com.cloud.vm.VirtualMachineManager;
+
+public class StorageFsVmFileShareLifeCycle implements FileShareLifeCycle {
+    protected Logger logger = LogManager.getLogger(getClass());
+
+    @Inject
+    private AccountManager accountMgr;
+
+    @Inject
+    protected ResourceManager resourceMgr;
+
+    @Inject
+    private VirtualMachineManager virtualMachineManager;
+
+    @Inject
+    private VolumeApiService volumeApiService;
+
+    @Inject
+    protected UserVmService userVmService;
+
+    @Inject
+    protected UserVmManager userVmManager;
+
+    @Inject
+    private DataCenterDao dataCenterDao;
+
+    @Inject
+    private VMTemplateDao templateDao;
+
+    @Inject
+    VolumeDao volumeDao;
+
+    @Inject
+    private UserVmDao userVmDao;
+
+    @Inject
+    NicDao nicDao;
+
+    @Inject
+    ServiceOfferingDao serviceOfferingDao;
+
+    @Inject
+    protected LaunchPermissionDao launchPermissionDao;
+
+    private String readResourceFile(String resource) {
+        try {
+            return FileUtil.readResourceFile(resource);
+        } catch (IOException e) {
+            throw new CloudRuntimeException("Unable to read the user data resource file due to exception " + e.getMessage());
+        }
+    }
+
+    private String getStorageFsVmConfig(final String fileSystem, final String hypervisorType) {
+        String fsVmConfig = readResourceFile("/conf/fsvm-init.yml");
+        final String filesystem = "{{ fsvm.filesystem }}";
+        final String hypervisor = "{{ fsvm.hypervisor }}";
+        fsVmConfig = fsVmConfig.replace(filesystem, fileSystem);
+        fsVmConfig = fsVmConfig.replace(hypervisor, hypervisorType);
+        return fsVmConfig;
+    }
+
+    private String getStorageFsVmName(String fileShareName) {
+        String prefix = String.format("%s-%s", FileShareVmNamePrefix, fileShareName);
+        String suffix = Long.toHexString(System.currentTimeMillis());
+
+        if (!NetUtils.verifyDomainNameLabel(prefix, true)) {
+            prefix = prefix.replaceAll("[^a-zA-Z0-9-]", "");
+        }
+        int nameLength = prefix.length() + suffix.length() + FileShareVmNamePrefix.length();
+        if (nameLength > 63) {
+            int prefixLength = prefix.length() - (nameLength - 63);
+            prefix = prefix.substring(0, prefixLength);
+        }
+        return (String.format("%s-%s", prefix, suffix));
+    }
+
+    private UserVm createFileShareVM(Long zoneId, Account owner, List<Long> networkIds, String name, Long serviceOfferingId, Long diskOfferingId, FileShare.FileSystemType fileSystem, Long size, Long minIops, Long maxIops) throws ResourceUnavailableException, InsufficientCapacityException, ResourceAllocationException {
+        ServiceOffering serviceOffering = serviceOfferingDao.findById(serviceOfferingId);
+
+        DataCenter zone = dataCenterDao.findById(zoneId);
+        Hypervisor.HypervisorType availableHypervisor = resourceMgr.getAvailableHypervisor(zoneId);
```

Comment:
@abh1sar can you please check behaviour with a multi-hypervisor env where the default hypervisor gets out of capacity?

I feel we should
- get the list of all available hypervisors and iterate over them
- check if the required template is present
- check if a deployment plan is possible

Reply:
Done.